core/domain/exp_domain.py

```diff
@@ -474,6 +474,246 @@ def __init__(self, change_list):
+            version_end: int. The upper bound of versions to get (exclusive).
+
+        Returns:
+            list(str). The names corresponding to the given range of versions.
+        """
+        return [
+            name for name, unused_state in (
+                self._get_multi_versions(version_start, version_end))
+        ]
+
+    def extend_or_split(self, exp_version, state_name, state, exp_version_diff):
+        """Extends span to include the given state's version if it is compatible
+        with previous versions, otherwise split it into its own distinct span.
+
+        Args:
+            exp_version: int. The state's corresponding exploration version.
+            state_name: str. The state's name at the given version.
+            state: state_domain.State. State's representation at given version.
+            exp_version_diff: ExplorationVersionsDiff. The changes since the
+                last exploration version.
+
+        Returns:
+            StateVersionSpan. The span chosen to hold the state.
+
+        Raises:
+            Exception. When any of the following occurs:
+                The state is different.
+                The given version has already been recorded.
+                The given version is being added out-of-sequence.
+        """
+        if exp_version != self._version_end:
+            raise Exception(
+                'Adding version=%d out-of-order (version=%d should be next)' % (
+                    exp_version, self._version_end))
+
+        prev_exp_version, prev_state_name = (
+            exp_version - 1,
+            exp_version_diff.new_to_old_state_names.get(state_name, state_name))
+
+        prev_snapshot_state_name, prev_snapshot_state = self[prev_exp_version]
+
+        if prev_state_name != prev_snapshot_state_name:
+            raise Exception(
+                'State does not map to previous version (previous name of '
+                'state is different; input exploration diff thinks that the '
+                'previous name was %r, but this span thinks that the previous '
+                'name was %r)' % (prev_state_name, prev_snapshot_state_name))
+
+        if prev_snapshot_state.is_structurally_compatible_with(state):
+            self._version_snapshots[exp_version] = (state_name, state)
+            self._version_end += 1
+            return self
+        else:
+            return StateVersionSpan(exp_version, state_name, state)
+
+    def _get_multi_versions(self, version_start, version_end):
+        """Returns state details for a half-open range of exploration versions.
+
+        Args:
+            version_start: int. The lower bound of versions to get (inclusive).
+            version_end: int. The upper bound of versions to get (exclusive).
+
+        Returns:
+            list(tuple(str, state_domain.State)). State details corresponding to
+            the given range of versions.
+        """
+        version_range = python_utils.RANGE(
```

Comment:
Perhaps better to error if the inputs are outside the valid range?

Reply:
Hmm but emulating Python's range API seems more idiomatic to me.

Reply:
I don't think this is comparable -- we have additional context here, and FWIW the idea of a comparison with Python's range() never occurred to me while reading this.

Reply:
Sorry, meant slices not range.

Reply:
Same response, though. The idea of a comparison with slice() never occurred to me, and I would not expect the two to be comparable a priori due to the differing contexts.

Reply:
Done.